### PR TITLE
Remove parsing literals of unrepresentable `SyntaxShape`s

### PR DESCRIPTION
**Pull request by sholderbach**
# Description
Those `SyntaxShape`s can not coerce to `Value`s or `Type`s that can be
used by the user in an argument or input-output-type position.
Supporting them doesn't make sense.

Open questions:
- Are cell-paths passable?
  - If so, should users make a distinction between full-.../cell-path
- Are errors passable (short-circuiting in the pipeline evaluation)?

# User-Facing Changes
Removal of useless "types" in argument type or input/output type positions

# Tests + Formatting
No adjustment necessary

### CODE REVIEW

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -2717,34 +2717,23 @@ pub fn parse_shape_name(
...
         b"float" => SyntaxShape::Float,
         b"filesize" => SyntaxShape::Filesize,
-        b"full-cell-path" => SyntaxShape::FullCellPath,
+        b"full-cell-path" => SyntaxShape::FullCellPath, // TODO: Should users distinguish cellpath?
         b"glob" => SyntaxShape::GlobPattern,
         b"int" => SyntaxShape::Int,
-        b"import-pattern" => SyntaxShape::ImportPattern,
-        b"keyword" => SyntaxShape::Keyword(vec![], Box::new(SyntaxShape::Any)),
         _ if bytes.starts_with(b"list") => parse_list_shape(working_set, bytes, span),
-        b"math" => SyntaxShape::MathExpression,
```
Comment (amtoine) [inline crates/nu-parser/src/parser.rs:2735]: a question that applies to all the others being removed:

if we do not use such a shape, could we remove it from the code base entirely? e.g. could we remove `SyntaxShape::MathExpression`? :smirk:

File: crates/nu-parser/src/parser.rs [hunk 2]
```diff
@@ -2717,34 +2717,23 @@ pub fn parse_shape_name(
...
-        // b"custom" => SyntaxShape::Custom(Box::new(SyntaxShape::Any), SyntaxShape::Int),
         b"datetime" => SyntaxShape::DateTime,
         b"directory" => SyntaxShape::Directory,
         b"duration" => SyntaxShape::Duration,
         b"error" => SyntaxShape::Error,
-        b"expr" => SyntaxShape::Expression,
         b"float" => SyntaxShape::Float,
         b"filesize" => SyntaxShape::Filesize,
-        b"full-cell-path" => SyntaxShape::FullCellPath,
+        b"full-cell-path" => SyntaxShape::FullCellPath, // TODO: Should users distinguish cellpath?
```
Comment (amtoine) [inline crates/nu-parser/src/parser.rs:2726]: tbh, i have no idea what a *full cell path* is as opposed to a regular *cell path* :thinking: 

one can do
```nushell
def foo [x: cell-path] { $x }
```
:+1:

**Comment by amtoine** [comment]
looking at all the shapes, i'm confused by the following three :eyes:
```nushell
> let p: path = "foo"; $p | describe
string
```
```nushell
> let g: glob = "*"; $g | describe
string
```
```nushell
> let d: directory = "foo"; $d | describe
string
```

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (sholderbach) [reply crates/nu-parser/src/parser.rs:2735]: `SyntaxShape` does more than `Type` and `MathExpression` tells the parser to parse a math expression. While useless for custom commands it is used by the `let`/`mut`/`const` "commands" to properly parse the assignment.

**Comment by sholderbach** [comment]
> looking at all the shapes, i'm confused by the following three 👀
> 
> ```nushell
> > let p: path = "foo"; $p | describe
> string
> ```
> 
> ```nushell
> > let g: glob = "*"; $g | describe
> string
> ```
> 
> ```nushell
> > let d: directory = "foo"; $d | describe
> string
> ```


This is one of the crimes of the "type-directed parsing" in Nushell that is more accurately described as "`SyntaxShape`-directed parsing". To enable the completions to do different things they can leverage the `SyntaxShape`. The parser is also free to do different things based upon those. But we only have `Value::String` and thus `Type::String` to represent those.

File: crates/nu-parser/src/parser.rs (hunk 2 above)
Comment (sholderbach) [reply crates/nu-parser/src/parser.rs:2726]: I think here I have to defer to @jntrnr on what the distinction is. https://github.com/nushell/nushell/pull/8493 introduced a literal syntax so cell-paths can easily be directly assigned and stored in their `Value` form.

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (amtoine) [reply crates/nu-parser/src/parser.rs:2735]: ooh ok gotcha, so let's keep them for internals :relieved:

**Comment by amtoine** [comment]
> But we only have Value::String and thus Type::String to represent those.
>
i feel it would make sense to have those too, right?

File: crates/nu-parser/src/parser.rs (hunk 2 above)
Comment (jntrnr) [reply crates/nu-parser/src/parser.rs:2726]: I think we could remove full-cell-path also as it'd end up just being evaluated before it could be used.

**Comment by fdncred** [comment]
You may want to put some comments in here explaining why all options do not exist. I usually find this stuff, and being a completest, come back in and add everything back.

**Pull request merged**